### ISSUE TITLE
NOPICKER not working

### ISSUE THREAD

**Issue by a10kiloham** [sickcodes/Docker-OSX#246]
I'm up and stable and using the naked image to persist through reboots. Everything works great except the picker always comes up. My Launch_custom.sh and Launch.sh both seem to be showing NOPICKER=true (and in the terminal I can see that env is set) but the picker seems to be confused. I always get the picker and have to click my image to start it. This probably a silly question but where does the QEMU get the image name from? Thanks!

Ssh'ing into the machine I see the Launch_custom lines as follows:
```
-drive id=OpenCoreBoot,if=none,snapshot=on,format=qcow2,file=${BOOTDISK:-/home/arch/OSX-KVM/OpenCore-Catalina/OpenCore.qcow2} \
-device ide-hd,bus=sata.2,drive=OpenCoreBoot \
-drive id=MacHDD,if=none,file=${IMAGE_PATH:-/home/arch/OSX-KVM/mac_hdd_ng.img},format=${IMAGE_FORMAT:-qcow2} \
```
Is the BOOTDISK variable being passed in incorrectly from the Dockerfile?

Dockerfile (customized to add VNC in)
```
FROM sickcodes/docker-osx:latest

MAINTAINER 'https://twitter.com/sickcodes' <https://sick.codes>

USER root

WORKDIR /root

RUN rm -f /home/arch/OSX-KVM/mac_hdd_ng.img

# OPTIONAL: Arch Linux server mirrors for super fast builds
# set RANKMIRRORS to any value other that nothing, e.g. -e RANKMIRRORS=true
ARG RANKMIRRORS
ARG MIRROR_COUNTRY=US
ARG MIRROR_COUNT=10
RUN if [[ "${RANKMIRRORS}" ]]; then { pacman -Sy wget --noconfirm || pacman -Syu wget --noconfirm ; } \
    ; wget -O ./rankmirrors "https://raw.githubusercontent.com/sickcodes/Docker-OSX/master/rankmirrors" \
    ; wget -O- "https://www.archlinux.org/mirrorlist/?country=${MIRROR_COUNTRY:-US}&protocol=https&use_mirror_status=on"
 \
    | sed -e 's/^#Server/Server/' -e '/^#/d' \
    | head -n "$((${MIRROR_COUNT:-10}+1))" \
    | bash ./rankmirrors --verbose --max-time 5 - > /etc/pacman.d/mirrorlist \
    && tee -a /etc/pacman.d/mirrorlist <<< 'Server = http://mirrors.evowise.com/archlinux/$repo/os/$arch' \
    && tee -a /etc/pacman.d/mirrorlist <<< 'Server = http://mirror.rackspace.com/archlinux/$repo/os/$arch' \
    && tee -a /etc/pacman.d/mirrorlist <<< 'Server = https://mirror.rackspace.com/archlinux/$repo/os/$arch' \
# TEMP-FIX for pacman issue
RUN patched_glibc=glibc-linux4-2.33-4-x86_64.pkg.tar.zst \
    && curl -LO "https://raw.githubusercontent.com/sickcodes/Docker-OSX/master/${patched_glibc}" \
    && bsdtar -C / -xvf "${patched_glibc}" || echo "Everything is fine."
# TEMP-FIX for pacman issue

# For taking screenshots of the Xfvb screen, useful during development.
ARG SCROT

RUN pacman -Syu xorg-server-xvfb wget xterm xorg-xhost xorg-xrandr sshpass --noconfirm \
    && if [[ "${SCROT}" ]]; then \
        pacman -Syu scrot base-devel --noconfirm \
        && git clone --recurse-submodules --depth 1 https://github.com/stolk/imcat.git \
        && cd imcat \
        && make \
        && sudo cp imcat /usr/bin/imcat \
        && touch /usr/bin/scrotcat \
        && tee -a /usr/bin/scrotcat <<< '/usr/bin/imcat <(scrot -o /dev/stdout)' \
        && chmod +x /usr/bin/scrotcat \
    ; else \
        touch /usr/bin/scrotcat \
        && echo echo >> /usr/bin/scrotcat \
        && chmod +x /usr/bin/scrotcat \
    ; fi \
    ; yes | pacman -Scc

# TEMP-FIX for pacman issue
RUN patched_glibc=glibc-linux4-2.33-4-x86_64.pkg.tar.zst \
USER arch

WORKDIR /home/arch/OSX-KVM

RUN mkdir -p ~/.ssh \
    && touch ~/.ssh/authorized_keys \
    && touch ~/.ssh/config \
    && chmod 700 ~/.ssh \
    && chmod 600 ~/.ssh/config \
    && chmod 600 ~/.ssh/authorized_keys \
    && tee -a ~/.ssh/config <<< 'Host *' \
    && tee -a ~/.ssh/config <<< '    StrictHostKeyChecking no' \
    && tee -a ~/.ssh/config <<< '    UserKnownHostsFile=/dev/null'

#### SPECIAL RUNTIME ARGUMENTS BELOW

ENV ADDITIONAL_PORTS=

ENV BOOTDISK=

ENV DISPLAY=:99

ENV HEADLESS=false

ENV ENV=/env
# Boolean for generating a bootdisk with new random serials.
ENV GENERATE_UNIQUE=false

# Boolean for generating a bootdisk with specific serials.
ENV GENERATE_SPECIFIC=false

ENV IMAGE_PATH=/image
ENV IMAGE_FORMAT=qcow2

ENV KVM='accel=kvm:tcg'

# ENV MASTER_PLIST_URL="https://raw.githubusercontent.com/sickcodes/osx-serial-generator/master/config-custom.plist"

# ENV NETWORKING=e1000-82545em
ENV NETWORKING=vmxnet3

ENV NOPICKER=true

# dynamic RAM options for runtime
ENV RAM=8
# ENV RAM=max
# ENV RAM=half

# The x and y coordinates for resolution.
# Must be used with either -e GENERATE_UNIQUE=true or -e GENERATE_SPECIFIC=true.
ENV WIDTH=1920
ENV HEIGHT=1080
CMD sudo touch /dev/kvm /dev/snd "${IMAGE_PATH}" "${BOOTDISK}" "${ENV}" || true \
    ; sudo chown -R $(id -u):$(id -g) /dev/kvm /dev/snd "${IMAGE_PATH}" "${BOOTDISK}" "${ENV}" || true \
    ; { [[ "${DISPLAY}" = ':99' ]] || [[ "${HEADLESS}" == true ]] ; } && { \
        nohup Xvfb :99 -screen 0 1920x1080x16 \
        & until [[ "$(xrandr --query 2>/dev/null)" ]]; do sleep 1 ; done \
    ; } \
    ; [[ "${NOPICKER}" == true ]] && { \
        sed -i '/^.*InstallMedia.*/d' Launch.sh \
        && export BOOTDISK="${BOOTDISK:=/home/arch/OSX-KVM/OpenCore-Catalina/OpenCore-nopicker.qcow2}" \
    ; } \
    || export BOOTDISK="${BOOTDISK:=/home/arch/OSX-KVM/OpenCore-Catalina/OpenCore.qcow2}" \
    ; [[ "${GENERATE_UNIQUE}" == true ]] && { \
        ./Docker-OSX/osx-serial-generator/generate-unique-machine-values.sh \
            --master-plist-url="${MASTER_PLIST_URL}" \
            --count 1 \
            --tsv ./serial.tsv \
            --bootdisks \
            --width "${WIDTH:-1920}" \
            --height "${HEIGHT:-1080}" \
            --output-bootdisk "${BOOTDISK:=/home/arch/OSX-KVM/OpenCore-Catalina/OpenCore.qcow2}" \
            --output-env "${ENV:=/env}" \
    ; } \
    ; [[ "${GENERATE_SPECIFIC}" == true ]] && { \
            source "${ENV:=/env}" 2>/dev/null \
            ; ./Docker-OSX/osx-serial-generator/generate-specific-bootdisk.sh \
            --master-plist-url="${MASTER_PLIST_URL}" \
            --model "${DEVICE_MODEL}" \
            --serial "${SERIAL}" \
            --board-serial "${BOARD_SERIAL}" \
            --uuid "${UUID}" \
            --mac-address "${MAC_ADDRESS}" \
            --width "${WIDTH:-1920}" \
            --height "${HEIGHT:-1080}" \
            --output-bootdisk "${BOOTDISK:=/home/arch/OSX-KVM/OpenCore-Catalina/OpenCore.qcow2}" \
    ; } \
    # ; ./enable-ssh.sh && /bin/bash -c ./Launch.sh

WORKDIR /home/arch/OSX-KVM

RUN sudo pacman -Syyuu --noconfirm \
    && sudo pacman -S tigervnc xterm xorg-xhost xdotool ufw --noconfirm \
    && mkdir -p ${HOME}/.vnc \
    && touch ~/.vnc/config \
    && tee -a ~/.vnc/config <<< 'geometry=1920x1080' \
    && tee -a ~/.vnc/config <<< 'localhost' \
    && tee -a ~/.vnc/config <<< 'alwaysshared'

RUN printf '\n%s\n' \
'sudo rm -f /tmp/.X99-lock' \
'export DISPLAY=:99' \
'/usr/bin/Xvnc -geometry 1920x1080 -rfbauth "${HOME}/.vnc/passwd" :99 &' > vnc.sh

RUN cat vnc.sh Launch.sh > Launch_custom.sh

RUN chmod +x Launch_custom.sh

RUN tee vncpasswd_file <<< "${VNC_PASSWORD:="$(tr -dc '[:graph:]' </dev/urandom | head -c8)"}"
RUN vncpasswd -f < vncpasswd_file > ${HOME}/.vnc/passwd

RUN chmod 600 ~/.vnc/passwd
CMD ./enable-ssh.sh && envsubst < ./Launch_custom.sh | bash
```

```
Linux macmini 5.4.0-71-generic #79-Ubuntu SMP Wed Mar 24 10:56:57 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
:0.0
1
NAME="Ubuntu"
PRETTY_NAME="Ubuntu 20.04.2 LTS"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
Filesystem      Size  Used Avail Use% Mounted on
/dev/nvme0n1p2  439G  242G  175G  58% /
QEMU emulator version 4.2.1 (Debian 1:4.2-3ubuntu6.14)
Copyright (c) 2003-2019 Fabrice Bellard and the QEMU Project developers
libvirtd (libvirt) 6.0.0
              total        used        free      shared  buff/cache   available
Mem:           62Gi       9.7Gi       489Mi        35Mi        52Gi        52Gi
Swap:         2.0Gi       2.0Mi       2.0Gi
8
8
crw-rw----+ 1 root kvm 10, 232 Apr 13 19:16 /dev/kvm
total 284K
drwxrwxrwt  2 root root 4.0K Apr 13 19:13 .
drwxrwxrwt 17 root root 276K Apr 14 09:55 ..
srwxrwxrwx  1 gdm  gdm     0 Apr 13 19:13 X0
root        1680  2.0  0.1 2939568 115476 ?      Ssl  Apr13  18:11 /usr/bin/dockerd -H fd:// --containerd=/run/containerd/containerd.sock
myname 108167  0.0  0.0   9036   664 pts/0    S+   09:55   0:00 grep --color=auto dockerd
docker:x:999:
kvm:x:133:
libvirt:x:140:myname
libvirt-qemu:x:64055:libvirt-qemu
libvirt-dnsmasq:x:141:
```

**Comment by a10kiloham** [comment]
Manually passing in the BOOTDISK variable works, so I'll just close this for now.

**Comment by sickcodes** [comment]
> Manually passing in the BOOTDISK variable works, so I'll just close this for now.

Thanks for pointing that out!

Do you think the Dockerfile is too abstract currently or is it good?

**Comment by a10kiloham** [comment]
It makes sense overall - I just thought it was a little hard to follow the flow since it pulls from the other repo so I was initially confused where the Launch.sh comes from. Might be useful to have a few prewritten Launch.sh scripts vs autogenerating them on the fly in the Dockerfiles as can be a bit more modular that way, but not a huge deal :)

**Comment by sickcodes** [comment]
> It makes sense overall - I just thought it was a little hard to follow the flow since it pulls from the other repo so I was initially confused where the Launch.sh comes from. Might be useful to have a few prewritten Launch.sh scripts vs autogenerating them on the fly in the Dockerfiles as can be a bit more modular that way, but not a huge deal :)

I agree, it's a bit messy at the moment, as we have two patches.

I saw you moved the disk stuff into the Launch script, I will consider making that normal :)

Just have to make sure it works again with `docker start` and all

**Comment by coppercash** [comment]
> Manually passing in the BOOTDISK variable works, so I'll just close this for now.

May I ask how did you "manually pass in the BOOTDISK variable" exactly?
`-e "BOOTDISK=/home/arch/OSX-KVM/OpenCore-Catalina/OpenCore-nopicker.qcow2"` doesn't work for me.

**Comment by sickcodes** [comment]
```
-e GENERATE_UNIQUE=true \
-e MASTER_PLIST_URL='https://raw.githubusercontent.com/sickcodes/osx-serial-generator/master/config-nopicker-custom.plist' \`
```
This will work

**Comment by coppercash** [comment]
Confirmed!
It works now!
Thanks a lot!
For others' reference, here is the complete command:

```bash
podman run -i --name mike \
    --device /dev/kvm \  # this line should be irrelevant
    -e EXTRA="-display none -vnc 0.0.0.0:99,password=on" \    # this line should be irrelevant
    -v `pwd`/volumes/image:/image \    # this line should be irrelevant
    -e SMP=30 -e CORES=1 -e RAM=56 \    # this line should be irrelevant
    -e WIDTH=1600 -e HEIGHT=900 \   # this line should be irrelevant
    -e NOPICKER=true \
    -e GENERATE_UNIQUE=true \
    -e MASTER_PLIST_URL='https://raw.githubusercontent.com/sickcodes/osx-serial-generator/master/config-nopicker-custom.plist' \
    --network=host \  # this line should be irrelevant
    docker.io/sickcodes/docker-osx:naked
```

**Comment by sickcodes** [comment]
`    -e NOPICKER=true \` is also irrelevant if you override it with your own `MASTER_PLIST_URL` which has nopicker on, because the full URL is: https://raw.githubusercontent.com/sickcodes/osx-serial-generator/master/config-nopicker-custom.plist

**Comment by coppercash** [comment]
> ` -e NOPICKER=true \` is also irrelevant if you override it with your own `MASTER_PLIST_URL` which has nopicker on, because the full URL is: https://raw.githubusercontent.com/sickcodes/osx-serial-generator/master/config-nopicker-custom.plist

Thanks. Great to know!
